### PR TITLE
Add `health_check` extension to bundled `otel.yml` file

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR adds the `health_check` extension to the `otel.yml` file that's bundled with the Elastic Agent package.

## Why is it important?

Without this addition, if a user tries to install the `open-telemetry/opentelemetry-collector` Helm chart with Elastic's distribution of the OTel Collector (i.e. Elastic Agent), the installation eventually fails.  Specifically, as noted by @andrzej-stencel in https://github.com/elastic/elastic-agent/issues/5092#issuecomment-2309613481:

> One problem I'm seeing when running the chart with this configuration is that the Health Check extension is not used in the `otel.yml` file, which results in Kubernetes restarting the collector container and eventually resulting in a `CrashLoopBackOff`. We should add the `health_check` extension to `service::extensions` in the `otel.yml` file that's bundled with the image.
 
## Checklist

- [ ] ~My code follows the style guidelines of this project~
- [ ] ~I have commented my code, particularly in hard-to-understand areas~
- [ ] ~I have made corresponding changes to the documentation~
- [x] I have made corresponding change to the default configuration files
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] ~I have added an integration test or an E2E test~

## Disruptive User Impact

If users are already running something that's listening on `localhost:13133`, there will be a port conflict.

## How to test this PR locally

See https://github.com/elastic/elastic-agent/issues/5092#issuecomment-2309509776 and https://github.com/elastic/elastic-agent/issues/5092#issuecomment-2309613481

### CODE REVIEW

File: otel.yml [hunk 1]
```diff
@@ -21,6 +21,9 @@ extensions:
     check_interval: 1s
     limit_mib: 700
     spike_limit_mib: 180
+  health_check:
+    endpoint: "localhost:13133"
```
Comment (ycombinator) [inline otel.yml:25]: I chose this value based on the documentation in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/healthcheckextension/README.md#health-check.  I'm happy to change it to another value if others have a better suggestion.

File: otel.yml [hunk 2]
```diff
@@ -21,6 +21,9 @@ extensions:
...
+    path: "/health/status"
```
Comment (ycombinator) [inline otel.yml:26]: The default value here, according to https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/healthcheckextension/README.md#health-check, is `/`.  But I think it's better to use a more specific path for the health check endpoint.  Again, happy to change this if others have a better suggestion.

File: otel.yml [hunk 3]
```diff
@@ -21,6 +21,9 @@ extensions:
...
 
 service:
   extensions: [memory_limiter]
```
Comment (rogercoll) [inline otel.yml:29]: Should we add it in the pipeline too?

```suggestion
  extensions: [health_check, memory_limiter]
```

File: otel.yml (hunk 2 above)
Comment (rogercoll) [reply otel.yml:26]: Maybe `/liveness`? In that sense, we will be more aligned to the `elastic-agent` health check endpoint https://github.com/elastic/elastic-agent/pull/4499

File: otel.yml (hunk 3 above)
Comment (ycombinator) [reply otel.yml:29]: Yes, duh 🤦 - thanks for catching!

File: otel.yml (hunk 2 above)
Comment (ycombinator) [reply otel.yml]: Done in 6f2306c6c8.  Thanks for the suggestion.

Comment (rogercoll) [reply otel.yml]: Just to give more context on the Helm configuration, any change on the default health_check configuration will need to be overrided in the `liveness and readinesss` proves of the Helm Chart config too:

https://github.com/open-telemetry/opentelemetry-helm-charts/blob/main/charts/opentelemetry-collector/values.yaml#L395

```yaml
livenessProbe:
  httpGet:
    port: 13133
    path: /health/status
readinessProbe:
  httpGet:
    port: 13133
    path: /health/status
```

**Comment by rogercoll** [comment]
Sounds good to me on adding the `health_check` extension in the bundled `otel.yml`. :+1: 

> Without this addition, if a user tries to install the open-telemetry/opentelemetry-collector Helm chart with Elastic's distribution of the OTel Collector (i.e. Elastic Agent), the installation eventually fails.

I think that this will only happen if the user tries to use the `otel.yml` file as it overrides the configuration set by the Chart:

```
$ kubectl describe pod  elastic-otelcol-opentelemetry-collector-agent-d9s4k
...
Command:
      //usr/share/elastic-agent/elastic-agent
    Args:
      --config=/conf/relay.yaml
      otel
      --config=otel.yml
```

Note that the Chart added another [default config](https://github.com/open-telemetry/opentelemetry-helm-charts/tree/main/charts/opentelemetry-collector#default-configuration) `--config=/conf/relay.yaml` (which contains the health_check extension). A workaround is to not use the `otel.yml` config and rely on the one configured by the Chart:

```
helm install elastic-otelcol open-telemetry/opentelemetry-collector --version=0.102.1 --set mode=daemonset --set image.repository="docker.elastic.co/beats/elastic-agent" --set image.tag="8.16.0-SNAPSHOT" --set image.pullPolicy=Always --set command.name="/usr/share/elastic-agent/elastic-agent" --set command.extraArgs="{otel}"
```

The recommened approach to provide a custom configuration for the collector is to manully create a [configuration configMap](https://github.com/elastic/opentelemetry-demo/blob/main/kubernetes/elastic-helm/deployment.yaml#L63). There is a PR upstream that will allow overriding the default configuration with a config option: https://github.com/open-telemetry/opentelemetry-helm-charts/pull/1301

File: otel.yml (hunk 2 above)
Comment (ycombinator) [reply otel.yml]: Thanks @rogercoll.  

Does this mean we will need to specify additional `--set` arguments to `helm chart install` for the configuration keys you mentioned above, assuming we are using the bundled `otel.yml` file via `--set command.name="/usr/share/elastic-agent/elastic-agent" --set command.extraArgs="{otel, --config=otel.yml}"`?  

So something like `--set livenessProbe.httpGet.path=/liveness --set readinessProbe.httpGet.path=/liveness`?

Sorry if this is a basic question but I'm a n00b when it comes to Helm.

Comment (rogercoll) [reply otel.yml]: Correct, you will need to override the probes path with the --set flag. But as I was explaining in this [comment](https://github.com/elastic/elastic-agent/pull/5369#issuecomment-2314505406), the upstream Helm Charts override the arguments launched within the container and specify their custom configuration args. That means that the `otel.yml` file will not be taken into consideration and a custom configuration should be provided with a K8s config map.

I am not 100% sure about changing the default `health_check` path in the sample configuration, maybe @elastic/elastic-agent-control-plane can share their feedback based on the current elastic-agent health endpoint.

**Review by rogercoll — APPROVED** [review]

File: otel.yml (hunk 2 above)
Comment (ycombinator) [reply otel.yml]: @blakerouse @andrzej-stencel @swiatekm do you have any opinions here?

Comment (blakerouse) [reply otel.yml]: The elastic-agent uses `/liveness` as the default endpoint for its own healthcheck. I don't know if we want to use the same for OTEL, as that would change the behavior of the default collector and probably don't want to do that.

I would like to also see v2 health check being used at some point, as it does a much better job of actually providing the true health of the collector.

Comment (ycombinator) [reply otel.yml]: Thanks @blakerouse. I will revert the `path` here in this PR to `/health/status` again for the reason you mentioned. 

Agreed on switching to healthcheckv2 in a future PR :)

**Review by blakerouse — APPROVED** [review]
Looks good, thanks for the updating the path.

**Pull request merged**